### ISSUE TITLE
Workflow: Build Own "Fork"

### ISSUE THREAD

**Issue by ax3l** [spack/spack#3104]
Currently, most packages in spack also have a `master`/`dev` "version" option that allows to get - without a fixed tag - the latest "head" of a software. First of all, this is awesome for developers.

My question is a bit directed towards quick contributions as we know it from GitHub/Bitbucket/etc.: within about 1minute or less, people are able to fork-&-fix an issue they see, be it documentation, a failing outdated tutorial example or a compile error. In most cases of such fixes, finding and fixing a very local problem, is easy while building & testing the much bigger project might actually be not so (e.g., because the user just took the latest stable release and developed against that).

In order to ease that workflow, do we somehow have an easy mean to e.g. build a branch from a self-forked package? So to say, can we express "build a certain *path/location* the same way as you build `@develop`"?

It would also ease user-support for application developers a lot, especially when third party software that an app depends on is broken (*looking at you `boost@master` used with `nvcc`*  :eyes: ).

I would love to tell our users and developers:
- install all dependencies of our app via spack
- user complains latest boost version is broken with our app
- we spot the problem and provide an upstream patch to boost
- user can in the meanwhile build against our patched boost and confirm it is working :1st_place_medal: 

Currently, I would need to suggest
- 1-3 as before
- *now build boost from our forked source by hand until our patch is accepted upstream* :cry: 

### TL;DR

```bash
spack install somelib
# while using the installed somelib: aww, it's not doing what I want
spack install somelib@develop
# while using the installed somelib: aww, even develop is broken

# GitHub: Fork & Patch!

# But how can I now quick-build my patched version?
# A)
spack install somelib@github.com:ax3l/somelib.git +branch=fix-someProblem
# B)
spack install somelib@develop +patch=fixSomeProblem.patch
# ...?
```

**Comment by davydden** [comment]
> do we somehow have an easy mean to e.g. build a branch from a self-forked package?

i think this is along the way of building a version of a `package.py` from some remote location. I think you can already do this with [package repositories](http://spack.readthedocs.io/en/latest/repositories.html?highlight=repository).

**Comment by davydden** [comment]
btw, i think it's not really comparable to versions of a package (such as `develop`), but rather a version of a `package.py`. 

So it's roughly
```
spack install path/to/git/fork.git/var/spack/repo/package.py
```

i think if you have additional checkout of your fork and add it as extra repositoreis, then this should work. It's not super neat, though.

**Comment by adamjstewart** [comment]
@davydden The only problem with that strategy is that you need to write your own package.py. It would be easier for the user to just hack the existing package.py and change the download repo and branch.

**Comment by davydden** [comment]
> change the download repo and branch.

that can be done with repositories, can't it?

**Comment by adamjstewart** [comment]
Yes, it just adds an extra step of creating a new repository with a new package. Really the only benefit is that you can hack the package outside of the builtin repo and not have to worry about it interfering with a push or pull since your package will differ from upstream.

**Comment by davydden** [comment]
> Yes, it just adds an extra step of creating a new repository with a new package.

that's not what i meant. I was thinking about the following workflow:

1. you keep you local `spack` up-to-date with your own fork, but use `develop` branch which is the same as upstream's develop
2. somewhere **outside** of the local spack, you checkout your GitHub fork again. But switch to your other branch, which has modified packages.
3. add (2) as remote repo.
4. One can also delete all un-modified packges on that branch, so that the remote repo will only have the needed packages.

Cumbersome, i now.

**Comment by ax3l** [comment]
thank you both for the quick feedback!

yes, I forgot to mention that writing an extra or hacking the existing spack package (if accessible) would be possible, similar as people write their own modules files on top of system module files, but it requires the spack user to leave the user CLI comfort zone and dig into the intrinsics of spack packages :)

**Comment by ax3l** [comment]
Maybe we could provide a way to add an "extra" version from command line?

E.g.
```bash
# I want to add "dev2" to the buildin libsplash package from cli
spack versions libsplash
# ==> Safe versions (already checksummed):
#   1.6.0  1.5.0  1.4.0  1.3.1  1.2.4  master  dev
# ==> Remote versions (not yet checksummed):
#   1.3.0  1.2.3  1.2.2  1.2.1  1.2.0

# version('dev2', branch='dev2',
#            git='https://github.com/ComputationalRadiationPhysics/libSplash.git')
spack versions add dev2 dev2@https://github.com/ComputationalRadiationPhysics/libSplash.git
```

**Comment by davydden** [comment]
homebrew makes such things possible with 
```
brew install username/repo/vim       # installs from your custom repo
```

> Maybe we could provide a way to add an "extra" version from command line?

if the only thing you need to change in the package is adding another versions or changing the `develop` to point elsewhere, i would just keep this modification in a local "stable" branch and merge upstream `develop` from time to time.

**Comment by ax3l** [comment]
I was thinking about just adding a new yaml db with additionally known versions in the user's `.spack/` per package. Syntax support as above and not more nor less information than what our python `version()` call provides.

**Comment by citibeth** [comment]
First of all, consider maintaining a Spack fork for your users.  Then you can push out new versions (whether checksummed or git), then tell people to `git pull` their Spack and install the new version.

You can also `spack diy` or `spack setup` any downloaded source tree, even if it's no version at all (i.e. not checked into Git).  Maybe this would be useful for your case?

**Comment by citibeth** [comment]
It's also possible to build Spack-only "release versions" of a package based on its Git hash.  See:

https://spack.readthedocs.io/en/latest/workflows.html#upstream-bug-fixes

**Comment by ax3l** [comment]
> First of all, consider maintaining a Spack fork for your users. Then you can push out new versions (whether checksummed or git), then tell people to git pull their Spack and install the new version.

If I would be a sysadmin, responsible for a finite number of HPC systems and setting up specific [queues and architectures](https://github.com/LLNL/spack/issues/2379) for that this would absolutely be the way to go. (I am also doing that as an early adopter user for our home cluster.)

But let me set up my "application developer hat" :tophat: in which I am only concerned about developing libraries and apps and letting users use it. In that case, systems will (hopefully) already have a specific forked version of spack running and telling my users not to use that one in their environment but to got for my outdated fork of spack instead is definitely not ideal.

> You can also spack diy or spack setup

sounds interesting. are there already docs for it? couldn't find more then the [help output and api reference](https://spack.readthedocs.io/en/latest/search.html?q=diy&check_keywords=yes&area=default) yet...

> It's also possible to build Spack-only "release versions" of a package based on its Git hash

That is what I am describing above and I want to spare the user from modifying or forking the spack package if the only thing that changes is a hand-selected new version.

**Comment by alalazo** [comment]
> sounds interesting. are there already docs for it? couldn't find more then the help output and api reference  yet...

Just FYI, there is an on-going effort to unify / homogeneize the support for software development in spack. You can find notes about it [here](https://github.com/LLNL/spack/wiki/Spack-Environments-WG).

**Comment by citibeth** [comment]
@alalazo I've come to believe that Spack Environments will be useful for a lot more than software dev.

> Did you mean "run your own package repository" instead?

You are right about the sysadmin vs. software dev roles, and maintaining a fork as a software dev is less practical.  I agree, running your own repo would be appropriate here.

> `spack diy` sounds interesting. are there already docs for it? 

Believe it or not, there are not docs.  Sorry for the oversight; see #3116

> `spack setup` is a way to go, I then have to ship a package.py to the user or recommend adding "my" package repository, overwrite the default third party I do not maintain usually and let the user use `spack install` as always.

Huh?  Not sure what you're saying here.  The prerequisites for `spack setup` (i.e. existence of a `package.py`) are the same as for `spack install` and `spack diy`.

> > It's also possible to build Spack-only "release versions" of a package based on its Git hash
> That is what I am [describing above](https://github.com/LLNL/spack/issues/3104#issuecomment-278654498) and I want to spare the user from modifying or forking the spack package if the only thing that changes is a hand-selected new version.

I think I see a new feature here.  But did want to first make sure that existing features do not suffice.  How about this...?

In addition to `url=`, we allow `git=` (or equivalent) to be added to the package, giving the standard git (or hg/svn/etc) URL for this package.  Spack should then be able to install any Git-downloaded version on-the-fly using an extra flag on git:
```
spack install mypackage@1.5.4.3.2 --git bugfix-branch
```
This would be roughly equivalent to the following that is possible today:
```
git clone git:/...mypackage.... -b bugfix-branch
cd mypackage
spack diy mypackage@1.5.4.3.2
```

**Comment by becker33** [comment]
@ax3l what do you think of the following:

```
packages:
  boost:
    package_attributes:
      git: 'git://github.com:my_fork/boost.git'
```

We've gotten this request from other users as well, and that was one proposal for how to resolve it. It's not quite as streamlined as doing it on the command line, but you can do it from the CLI

```
spack config add packages:boost:package_attributes:git:"git://github.com:my_fork/boost.git"
spack install foo^boost@git.my_fix_branch
```

or even as a one-liner
```
spack -c packages:boost:package_attributes:git:"git://github.com:my_fork/boost.git" install foo^boost@git.my_fix_branch
```

**Comment by Matthew-Whitlock** [comment]
It would be convenient if the package_attributes config could be conditionally applied - e.g. a "when: @git.my_fix_branch" at some related scope.

Another convenient place to put this information would be in `spack develop`. Something like 
```
spack develop --path ./my_boost_src --git git://github.com:my_fork/boost.git boost@git.my_fix_branch
```
And then in the spack.yaml for an environment, you could have 
```
develop:
  boost:
    spec: boost@git.my_fix_branch
    path: my_boost_src
    #Either:
    git: 'git://github.com:my_fork/boost.git'
    # Or for more general support:
    package_attributes:
      git: 'git://github.com:my_fork/boost.git'
```

**Comment by becker33** [comment]
Closing as completed with #33275 

@Matthew-Whitlock if you are still interested in conditional `package_attributes` overrides, please open a separate issue for that
